### PR TITLE
Refactor README.md for PGXN

### PR DESCRIPTION
**Pull request by mkgrgis**
Refactor to advanced FDW documentation README.md structure implemented in
- **https://github.com/jeffreydwalter/redis_cluster_fdw**
- https://github.com/ibarwick/firebird_fdw
- https://github.com/pgspider/sqlite_fdw
- https://github.com/pgspider/jdbc_fdw
- https://github.com/pgspider/influxdb_fdw
- https://github.com/pgspider/dynamodb_fdw
- https://github.com/pgspider/griddb_fdw

Reorder options in the main fdw file.

### CODE REVIEW

**Comment by mkgrgis** [comment]
Hello, @adunstan ! Have you got a time to review unified `REDDME.md` about your FDW for PGXN?

File: README.md [hunk 1]
```diff
@@ -1,105 +1,152 @@
...
-==============================
+Redis Foreign Data Wrapper for PostgreSQL
+==========================================
 
-This PostgreSQL extension implements a Foreign Data Wrapper (FDW) for
-the Redis key/value database: http://redis.io/
+This is a foreign data wrapper (FDW) to connect [PostgreSQL](https://www.postgresql.org/)
+to [Redis](http://redis.io/) key/value database. This FDW works with PostgreSQL 10+ and confirmed with some Redis versions near 6.0.
+
+<img src="https://upload.wikimedia.org/wikipedia/commons/2/29/Postgresql_elephant.svg" align="center" height="100" alt="PostgreSQL"/>	+	<img src="https://avatars.githubusercontent.com/u/1529926" align="center" height="100" alt="Redis"/>
```
Comment (dpage) [inline README.md:7]: Don't link to images etc. on 3rd party sites as it gives those sites the IP address of visitors to our resources.

File: redis_fdw.c
Comment (dpage) [inline redis_fdw.c:1]: This is unrelated to a README restructuring. Please remove it from this PR and submit separately.

File: README.md [hunk 2]
```diff
@@ -1,105 +1,152 @@
...
 
-- Redis has acquired cursors as of Release 2.8. This is used in all the
-  mainline branches from REL9_2_STABLE on, for operations which would otherwise
-  either scan the entire Redis database in a single sweep, or scan a single,
-  possible large, keyset in a single sweep. Redis Releases prior to 2.8 are
-  maintained on the REL9_x_STABLE_pre2.8 branches.
+#### Build and install on OS
 
-- Redis cursors have some significant limitations. The Redis docs say:
+Add a directory of `pg_config` to PATH and build and install `regis_fdw`.
```
Comment (dpage) [inline README.md:81]: The phrasing here needs to be reworked as it only just makes sense as written, and may be difficult for some to understand. Also, redid_fdw is misspelt.

File: README.md [hunk 3]
```diff
@@ -108,85 +155,239 @@ in the case of lists sets and scalars, rows with key and value text columns
...
+`redis_fdw` **doesn's support** [IMPORT FOREIGN SCHEMA](https://www.postgresql.org/docs/current/sql-importforeignschema.html) and accepts no custom options for this command. There is no formal storing schema in Redis in oppose to RDBMS.
 
-Insert, Update and Delete
--------------------------
+## TRUNCATE support
 
-PostgreSQL acquired support for modifying foreign tables in release 9.3, and
-now the Redis Foreign Data Wrapper supports these too, for 9.3 and later
-PostgreSQL releases. There are a few restriction on this:
+`redis_fdw` don't implements the foreign data wrapper `TRUNCATE` API, available
```
Comment (dpage) [inline README.md:164]: s/don't/doesn't/

File: README.md [hunk 4]
```diff
@@ -108,85 +155,239 @@ in the case of lists sets and scalars, rows with key and value text columns
 for hashes, and rows with a value text columns and an optional numeric score
 column for zsets.
 
-The following parameter can be set on a user mapping for a Redis
-foreign server:
+## IMPORT FOREIGN SCHEMA options
 
-password:	The password to authenticate to the Redis server with.
-     Default: <none>
+`redis_fdw` **doesn's support** [IMPORT FOREIGN SCHEMA](https://www.postgresql.org/docs/current/sql-importforeignschema.html) and accepts no custom options for this command. There is no formal storing schema in Redis in oppose to RDBMS.
```
Comment (dpage) [inline README.md:160]: s/doesn's/doesn't

File: README.md [hunk 5]
```diff
@@ -108,85 +155,239 @@ in the case of lists sets and scalars, rows with key and value text columns
...
+ - https://pgxn.org/tag/fdw/
+
+License and authors
+-------
+* Dave Page dpage@pgadmin.org
+* Andrew Dunstan andrew@dunslane.net
 
-Andrew Dunstan
-andrew@dunslane.net
+See the [`License`](License) file for full details.
```
Comment (dpage) [inline README.md:393]: There is no license file.

File: README.md [hunk 6]
```diff
@@ -108,85 +155,239 @@ in the case of lists sets and scalars, rows with key and value text columns
...
-	CREATE SERVER redis_server
-		FOREIGN DATA WRAPPER redis_fdw
-		OPTIONS (address '127.0.0.1', port '6379');
+PostgreSQL folds identifiers to lower case by default, Redis is case sensetive by default. It's important
+to be aware of potential issues with table and column names. If there will no proper name qouting in PostgreSQL, access from PostgreSQL foreign tables with mixedcase or uppercase names to mixedcase or uppercase Redis objects can cause unexpected results.
+
+Generated columns
+-----------------
+
+Redis doens't provides support for generated columns.
```
Comment (dpage) [inline README.md:182]: s/provides/provide/

**Review by dpage — CHANGES_REQUESTED** [review]
I've found various typos and missing words etc, however, I'm out of time for this for now. Please update per the comments I've added, and double-check for any more typos etc.

Thanks.

File: README.md (hunk 1 above)
Comment (mkgrgis) [reply README.md:7]: There are no web request redirect to this resources is we meant github hosting and md web interface. This is only image sources for GitHub external artefacts cache, please ensure and verify [this similar README](https://github.com/pgspider/sqlite_fdw). 
If camo.githubusercontent.com is not welcome or this repo is not for GitHub only, I can remove this links. What will better for you, @dpage ?

File: redis_fdw.c
Comment (mkgrgis) [reply redis_fdw.c:1]: Reverted in  https://github.com/pg-redis-fdw/redis_fdw/pull/37/commits/01bce9dfdce3ebcfe0dc6dc5e5ad05abede8315a, please verify @dpage !

File: README.md (hunk 2 above)
Comment (mkgrgis) [reply README.md]: Changed in https://github.com/pg-redis-fdw/redis_fdw/pull/37/commits/01bce9dfdce3ebcfe0dc6dc5e5ad05abede8315a, please review the new variant.

File: README.md (hunk 3 above)
Comment (mkgrgis) [reply README.md]: Fixed in https://github.com/pg-redis-fdw/redis_fdw/pull/37/commits/01bce9dfdce3ebcfe0dc6dc5e5ad05abede8315a, please verify.

File: README.md (hunk 4 above)
Comment (mkgrgis) [reply README.md]: Fixed, please ensure all is OK now.

File: README.md [hunk 7]
```diff
@@ -108,85 +155,239 @@ in the case of lists sets and scalars, rows with key and value text columns
...
+- [Generated Columns](https://www.postgresql.org/docs/current/ddl-generated-columns.html)
+- [CREATE FOREIGN TABLE](https://www.postgresql.org/docs/current/sql-createforeigntable.html)
 
-	CREATE FOREIGN TABLE myredishash (key text, val text[])
-		SERVER redis_server
-		OPTIONS (database '0', tabletype 'hash', tablekeyprefix 'mytable:');
+Character set handling
+----------------------
+
+All strings from Redis are interpreted acording the PostgreSQL database's server encoding. Redis supports UTF-8 only data.
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply README.md]: Changed in https://github.com/pg-redis-fdw/redis_fdw/pull/37/commits/01bce9dfdce3ebcfe0dc6dc5e5ad05abede8315a, please ensure all is OK here.

File: README.md (hunk 1 above)
Comment (dpage) [reply README.md:7]: I don't fully understand what you're asking. However, the image source specified is on wikimedia.org, which means that owners of that site will see traffic to this readme whenever it is rendered. Please include any images in the repo directly, and reference them relatively, so even if someone checks out the repo and then reads the README, there is no traffic from their system to Github or Wikimedia etc.

File: redis_fdw.c [hunk 8]
```diff
@@ -82,7 +82,6 @@ struct RedisFdwOption
  */
 static struct RedisFdwOption valid_options[] =
 {
-
```
Comment (dpage) [inline redis_fdw.c:85]: Please remove this unrelated whitespace change.

**Review by dpage — CHANGES_REQUESTED** [review]
I've made a couple of comments inline requesting specific changes. Thanks.

File: README.md (hunk 1 above)
Comment (mkgrgis) [reply README.md:7]: > However, the image source specified is on wikimedia.org, which means that owners of that site will see traffic to this readme whenever it is rendered

No. You can verify this fact. This means owners of this site sometimes have a traffic of GitHub outer artefacts cache, but not in moment of README rendering and not in every moment of a new rendering.

File: redis_fdw.c (hunk 8 above)
Comment (mkgrgis) [reply redis_fdw.c]: No problems, please verify, @dpage !

File: README.md (hunk 1 above)
Comment (dpage) [reply README.md:7]: Github may cache the asset and update the URL to render from their cache, but not everyone uses Github to render markdown. Anyone rendering on their local machine will almost certainly expose their IP address to Wikimedia.

Further, storing assets like that outside of the repo means that you cannot view the readme as intended if you're offline.

File: README.md [hunk 9]
```diff
@@ -108,85 +156,242 @@ in the case of lists sets and scalars, rows with key and value text columns
...
+ - https://pgxn.org/tag/fdw/
+
+License and authors
+-------
+* Dave Page dpage@pgadmin.org
+* Andrew Dunstan andrew@dunslane.net
 
-Andrew Dunstan
-andrew@dunslane.net
+There is no licese file here.
```
Comment (dpage) [inline README.md:397]: Please remove this line.

**Review by dpage — CHANGES_REQUESTED** [review]
Please see inline comments/discussion.

File: README.md (hunk 1 above)
Comment (mkgrgis) [reply README.md:7]: Ok, files will be added in the repo. Thanks, @dpage , I have understood your reason.

File: README.md (hunk 9 above)
Comment (mkgrgis) [reply README.md]: Please verify, changed in https://github.com/pg-redis-fdw/redis_fdw/pull/37/commits/90bac01fa56309fe174b8dc94dbcb1abcbb4200b

**Comment by mkgrgis** [comment]
@dpage , if all is ok now, please don't forget to squash my commits before merging.
